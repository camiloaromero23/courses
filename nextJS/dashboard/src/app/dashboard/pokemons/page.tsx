import { PokemonGrid, PokemonsResponse, SimplePokemon } from "@/pokemons";

export const metadata = {
  title: "Pokemons list",
  description: "Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
}

const getPokemons = async (
  limit = 20,
  offset = 0,
): Promise<SimplePokemon[]> => {
  const res = await fetch(
    `https://pokeapi.co/api/v2/pokemon?limit=${limit}&offset=${offset}`,
  );

  const data: PokemonsResponse = await res.json();

  const pokemons = data.results.map((pokemon) => ({
    id: pokemon.url.split("/").at(-2)!,
    name: pokemon.name,
  }));

  // throw new Error("Not implemented");
  // throw notFound();

  return pokemons;
};

export default async function PokemonsPage() {
  const pokemons = await getPokemons(150);

  return (
    <div className="flex flex-col p-2">
      <span className="text-5xl my-2">
        Pokemons list <small className="text-blue-500">static</small>
      </span>

      <PokemonGrid pokemons={pokemons} />
    </div>
  );
}
