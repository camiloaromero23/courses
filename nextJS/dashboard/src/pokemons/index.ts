export { PokemonGrid } from "./components/PokemonGrid";
export { FavoritePokemons } from "./components/FavoritePokemons";

export type { PokemonsResponse } from "./interfaces/pokemonsResponse";
export type { SimplePokemon } from "./interfaces/simplePokemon";
export type { Pokemon } from "./interfaces/pokemon";
