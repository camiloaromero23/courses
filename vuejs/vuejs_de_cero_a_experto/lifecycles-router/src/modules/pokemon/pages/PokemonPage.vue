<template>
  <h1>Pokemon #{{ id }}</h1>
  <div v-if="pokemon">
    <img :src="pokemon.sprites.front_default" :alt="pokemon.name" />
  </div>
</template>

<script>
export default {
  props: {
    id: {
      type: Number,
      required: true,
    },
  },
  data() {
    return {
      pokemon: null,
    };
  },
  created() {
    /* console.log(this.$route.params.id); */
    /* this.id = +this.$route.params.id; */
    this.getPokemon();
  },
  watch: {
    id() {
      this.getPokemon();
    },
  },
  methods: {
    async getPokemon() {
      try {
        const res = await fetch(`https://pokeapi.co/api/v2/pokemon/${this.id}`);
        const pokemon = await res.json();
        this.pokemon = pokemon;
        console.log(pokemon);
      } catch (error) {
        this.$router.push('/');
        console.log('Nothing left to do here.');
      }
    },
  },
};
</script>
