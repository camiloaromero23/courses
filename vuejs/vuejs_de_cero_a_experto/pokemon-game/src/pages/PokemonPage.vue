<template>
  <h1>Who's that pok√©mon?</h1>
  <h1 v-if="!pokemon">Please wait...</h1>
  <div v-else>
    <PokemonPicture
      :pokemonId="pokemon.id"
      :showPokemon="showPokemon"
    ></PokemonPicture>
    <PokemonOptions
      :optionsDisabled="!showResult"
      :pokemons="pokemons"
      @pokemonSelected="checkAnswer"
    ></PokemonOptions>
    <template v-if="showResult">
      <h2 class="fade-in">{{ message }}</h2>
      <button @click="newGame">New Game</button>
    </template>
  </div>
</template>

<script>
import PokemonPicture from '../components/PokemonPicture';
import PokemonOptions from '../components/PokemonOptions';
import getPokemonOptions from '../helpers/getPokemonOptions';

export default {
  components: {
    PokemonPicture,
    PokemonOptions,
  },
  data() {
    return {
      pokemons: [],
      pokemon: null,
      showPokemon: false,
      showResult: false,
      message: '',
    };
  },
  methods: {
    async mixPokemons() {
      this.pokemons = await getPokemonOptions();
      const randomInt = Math.floor(Math.random() * 4);
      this.pokemon = this.pokemons[randomInt];
    },
    toggleImage() {
      this.showPokemon = !this.showPokemon;
    },
    toggleShowResult() {
      this.showResult = !this.showResult;
    },
    checkAnswer(pokemonId) {
      this.toggleImage();
      if (pokemonId === this.pokemon.id) {
        this.message = 'Correct! ';
      } else {
        this.message = 'Whoops! The right one was ';
      }
      this.message += this.pokemon.name;
      this.toggleShowResult();
    },
    newGame() {
      this.toggleShowResult();
      this.toggleImage();
      this.mixPokemons();
      this.pokemons = [];
      this.pokemon = null;
    },
  },
  mounted() {
    this.mixPokemons();
  },
};
</script>

<style>
ul {
  list-style-type: none;
}
li {
  background-color: White;
  border-radius: 5px;
  border: 1px solid rgba(0, 0, 0, 0.2);
  cursor: pointer;
  margin-bottom: 10px;
  width: 250px;
}

li:hover {
  background-color: rgba(0, 0, 0, 0.05);
}

.options-container {
  display: flex;
  justify-content: center;
}
</style>
