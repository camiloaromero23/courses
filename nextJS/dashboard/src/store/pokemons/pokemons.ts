import { SimplePokemon } from "@/pokemons";
import { PayloadAction, createSlice } from "@reduxjs/toolkit";

export interface PokemonsState {
  favorites: {
    [key: string]: SimplePokemon;
  };
}

// const getInitialState = (): PokemonsState => {
//   // if (typeof localStorage === "undefined") return {}; // WARN: Messes with hydration

//   const pokemons = localStorage.getItem("pokemons");

//   return pokemons ? JSON.parse(pokemons) : {};
// };

const initialState: PokemonsState = {
  favorites: {}
  // ...getInitialState(),
  // "1": { id: "1", name: "bulbasaur" },
  // "3": { id: "3", name: "venusaur" },
  // "5": { id: "5", name: "charmeleon" },
};

const pokemonsSlice = createSlice({
  name: "pokemons",
  initialState,
  reducers: {
    setFavoritePokemons(state, action: PayloadAction<typeof initialState.favorites>) {
      state.favorites = action.payload;
    },

    toggleFavorite(state, action: PayloadAction<SimplePokemon>) {
      const pokemon = action.payload;

      if (!!state.favorites[pokemon.id]) {
        delete state.favorites[pokemon.id];
      } else {
        state.favorites[pokemon.id] = pokemon;
      }
    },
  },
});

export const { toggleFavorite, setFavoritePokemons } = pokemonsSlice.actions;
export default pokemonsSlice.reducer;
