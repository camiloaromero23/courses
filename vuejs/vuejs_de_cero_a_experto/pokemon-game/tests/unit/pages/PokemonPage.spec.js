import { shallowMount } from '@vue/test-utils';
import PokemonPage from '@/pages/PokemonPage';
import { pokemons } from '../mocks/pokemons.mock';

describe('PokemonPage Component', () => {
  let wrapper;
  let loadedWrapper;

  beforeEach(() => {
    wrapper = shallowMount(PokemonPage);
    loadedWrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showResult: false,
          message: '',
          selectedPokemon: null,
        };
      },
    });
  });
  it('should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should call mixPokemonArray once mounted', () => {
    const mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemons');
    shallowMount(PokemonPage);
    expect(mixPokemonArraySpy).toHaveBeenCalled();
  });

  it('should match snapshot when pokemons are loaded', () => {
    loadedWrapper;
    expect(loadedWrapper.html()).toMatchSnapshot();
  });

  it('should show PokemonPicture & PokemonOptions components', () => {
    const PokemonPicture = loadedWrapper.find('pokemon-picture-stub');
    const PokemonOptions = loadedWrapper.find('pokemon-options-stub');

    expect(PokemonPicture.exists()).toBeTruthy();
    expect(PokemonOptions.exists()).toBeTruthy();

    expect(PokemonPicture.attributes('pokemonid')).toBe('1');
    expect(PokemonOptions.attributes('pokemons')).toBeTruthy();
  });

  it('should update checkAnswer', async () => {
    await loadedWrapper.vm.checkAnswer(1);
    const h2Tag = loadedWrapper.find('h2');

    expect(h2Tag.exists()).toBeTruthy();
    expect(h2Tag.text()).toBe(`Correct! ${pokemons[0].name}`);
    expect(loadedWrapper.vm.showPokemon).toBeTruthy();
    await loadedWrapper.vm.checkAnswer(11);
    expect(loadedWrapper.vm.message).toBe(
      `Whoops! The right one was ${pokemons[0].name}`,
    );
  });
});
