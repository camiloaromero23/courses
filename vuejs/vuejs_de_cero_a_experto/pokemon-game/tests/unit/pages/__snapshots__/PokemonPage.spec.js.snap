// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component should match snapshot when pokemons are loaded 1`] = `
<h1>Who's that pokémon?</h1>
<div>
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]" pokemonid="1" optionsenabled="true"></pokemon-options-stub>
  <!--v-if-->
</div>
`;

exports[`PokemonPage Component should match the snapshot 1`] = `
<h1>Who's that pokémon?</h1>
<h1>Please wait...</h1>
`;
